dashboard/test/helpers/delete_accounts_helper_test.rb

```diff
@@ -0,0 +1,254 @@
+require 'test_helper'
+require 'cdo/delete_accounts_helper'
+
+#
+# This test is the comprehensive spec on the desired behavior when purging a
+# user from our system (that is, hard-deleting a user).  If you need to change
+# the user purge process, start by testing the desired behavior here, then work
+# your way down to the appropriate implementation point.
+#
+# Purging a user is an irreversible operation.  We don't delete every row that
+# has ever been associated with that user, but we do remove all sensitive
+# information, anything that might possibly be used to identify the particular
+# user, and anything we don't need to keep around for our metrics.
+#
+# Getting this right is important for compliance with various privacy
+# regulations around the globe, so changes to this behavior should be carefully
+# reviewed by the product team.
+#
+class DeleteAccountsHelperTest < ActionView::TestCase
+  test 'sets purged_at' do
+    user = create :student
+    assert_nil user.purged_at
+
+    purge_user user
+
+    user.reload
+    refute_nil user.purged_at
+  end
+
+  test 'clears user.name' do
+    user = create :student
+    refute_nil user.name
+
+    purge_user user
+
+    user.reload
+    assert_nil user.name
+  end
+
+  test 'anonymizes user.username' do
+    user = create :student
+    refute_nil user.username
+    refute_match /^sys_deleted_\w{8}$/, user.username
+
+    purge_user user
+
+    user.reload
+    assert_match /^sys_deleted_\w{8}$/, user.username
+  end
+
+  test 'clears user.*_sign_in_ip' do
+    user = create :student,
+      current_sign_in_ip: '192.168.0.1',
+      last_sign_in_ip: '10.0.0.1'
+    refute_nil user.current_sign_in_ip
+    refute_nil user.last_sign_in_ip
+
+    purge_user user
+
+    user.reload
+    assert_nil user.current_sign_in_ip
+    assert_nil user.last_sign_in_ip
+  end
+
+  test 'clears user email fields' do
+    user = create :teacher,
+      parent_email: 'fake-parent-email@example.com'
+    refute_nil user.email
+    refute_nil user.hashed_email
+    refute_nil user.parent_email
+
+    purge_user user
+
+    user.reload
+    assert_empty user.email
+    assert_empty user.hashed_email
+    assert_nil user.parent_email
```

Comment:
Out of curiosity, why is `parent_email` nil, and the others empty? Should we be consistent?

Also note that empty string will pass `refute_nil` so the initial 2 assertions don't necessarily show that `email` and `hashed_email` changed.

Reply:
Kind of a legacy thing - `email` has `not null` in our schema even though in practice it's often empty, but `parent_email`, added later, uses null as its empty value.

https://github.com/code-dot-org/code-dot-org/blob/132e640610c988b3a0c68773106d5dfc7ecd46b0/dashboard/app/models/user.rb#L7-L8

I'm in favor of consistency but think it's a concern that should be addressed in its own change, which would include updating these tests.

Thanks for catching the insufficient assertion here!